omx/gstomxvideofilter.c

```diff
@@ -0,0 +1,2398 @@
+      "omxvideofilter", 0, "OMX video filter base class");
+
+  parent_class = g_type_class_peek_parent (klass);
+
+  g_type_class_add_private (klass, sizeof (GstOMXVideoFilterPrivate));
+
+  gobject_class->finalize = gst_omx_video_filter_finalize;
+  gobject_class->set_property = gst_omx_video_filter_set_property;
+  gobject_class->get_property = gst_omx_video_filter_get_property;
+
+  g_object_class_install_property (gobject_class, PROP_ALWAYS_COPY,
+      g_param_spec_boolean ("always-copy", "Always copy",
+          "If the buffer will be used or not directly for the OpenMax component",
+          GST_OMX_VIDEO_FILTER_ALWAYS_COPY_DEFAULT, G_PARAM_WRITABLE));
+
+  g_object_class_install_property (gobject_class, PROP_OUTPUT_BUFFERS,
+      g_param_spec_uint ("output-buffers", "Output buffers",
+          "The amount of OMX output buffers",
+          1, 16, GST_OMX_VIDEO_FILTER_OUTPUT_BUFFERS_DEFAULT,
+          G_PARAM_READWRITE));
+
+  g_object_class_install_property (gobject_class, PROP_INPUT_BUFFERS,
+      g_param_spec_uint ("input-buffers", "Input buffers",
+          "The amount of OMX input buffers",
+          1, 16, GST_OMX_VIDEO_FILTER_INPUT_BUFFERS_DEFAULT,
+          G_PARAM_READWRITE));
+
+  element_class->change_state =
+      GST_DEBUG_FUNCPTR (gst_omx_video_filter_change_state);
+
+
+  klass->cdata.type = GST_OMX_COMPONENT_TYPE_FILTER;
+  klass->cdata.default_sink_template_caps = "video/x-raw, "
+      "width = " GST_VIDEO_SIZE_RANGE ", "
+      "height = " GST_VIDEO_SIZE_RANGE ", " "framerate = " GST_VIDEO_FPS_RANGE;
+
+  klass->transform_caps =
+      GST_DEBUG_FUNCPTR (gst_omx_video_filter_default_transform_caps);
+  klass->fixate_caps =
+      GST_DEBUG_FUNCPTR (gst_omx_video_filter_default_fixate_caps);
+}
+
+static void
+gst_omx_video_filter_reset (GstOMXVideoFilter * self)
+{
+  GstOMXVideoFilterPrivate *priv = self->priv;
+  GList *g;
+
+  GST_OMX_VIDEO_FILTER_STREAM_LOCK (self);
+
+  priv->drained = TRUE;
+
+  /* Free video frames from src pad pending list */
+  if (priv->frames) {
+    g_hash_table_foreach (priv->frames, (GHFunc) free_frame_array, NULL);
+    g_hash_table_destroy (priv->frames);
+  }
+  priv->frames = NULL;
+
+  for (g = priv->output_info; g; g = g->next) {
+    _free_video_info ((GstVideoInfo *) g->data);
+  }
+  g_list_free (priv->output_info);
+  priv->output_info = NULL;
+
+  gst_video_info_init (&priv->input_info);
+
+  GST_OMX_VIDEO_FILTER_STREAM_UNLOCK (self);
+}
+
+static void
+gst_omx_video_filter_init (GstOMXVideoFilter * self,
+    GstOMXVideoFilterClass * klass)
+{
+  GstOMXVideoFilterPrivate *priv;
+  GstPadTemplate *pad_template;
+  GstPad *pad;
+  gint i;
+
+  g_return_if_fail (klass->num_outputs > 0);
+
+  priv = self->priv = GST_OMX_VIDEO_FILTER_GET_PRIVATE (self);
+
+  pad_template =
+      gst_element_class_get_pad_template (GST_ELEMENT_CLASS (klass), "sink");
+  g_return_if_fail (pad_template != NULL);
+
+  self->sinkpad = pad = gst_pad_new_from_template (pad_template, "sink");
```

Comment:
Use the normal element definition, and use your pad templates.

Reply:
Pad templates need to be handled in cdata because of the logic present in gstomx.c to register the elements.